src/components/expression/expression-loader.js

```diff
@@ -24,101 +31,210 @@ genotet.utils.inherit(genotet.ExpressionLoader, genotet.ViewLoader);
-        }
-      });
-      geneRegex = geneRegex.slice(0, -1);
-      break;
-    case 'setCondition':
-      // Totally replace the regex.
-      conditionRegex = regex;
-      break;
-    case 'addCondition':
-      // Concat the two regex's. We need to include the previously existing
-      // genes too so as to find the edges between the new genes and old
-      // ones.
-      conditionRegex += '|' + regex;
-      break;
-    case 'removeCondition':
-      // Remove the regex.
-      conditionRegex = '';
-      this.data.matrix.conditionNames.forEach(function(conditionName) {
-        if (!conditionName.toUpperCase().match(regex)) {
-          conditionRegex += conditionName + '|';
-        }
-      });
-      conditionRegex = conditionRegex.slice(0, -1);
-      break;
-  }
-  this.load(matrixName, geneRegex, conditionRegex);
+                                                     names) {
+  var heatmapData = this.data.matrix;
+  var currentStatus = new genotet.ExpressionRenderer.ZoomStatus({
+    matrixName: heatmapData.matrixName,
+    dataName: heatmapData.dataName,
+    geneNames: heatmapData.geneNames,
+    conditionNames: heatmapData.conditionNames
+  });
+  this.data.zoomStack.push(currentStatus);
+  this.data.zoomStack.forEach(function(zoomStatus) {
+    switch (method) {
+      case 'setGene':
+        // Totally replace the names.
+        zoomStatus.geneNames = names;
+        break;
+      case 'addGene':
+        // Concat the two names.
+        this.removeNames_(zoomStatus.geneNames, names);
+        names.forEach(function(name) {
+          zoomStatus.geneNames.push(name);
+        });
+        break;
+      case 'removeGene':
+        // Remove the names.
+        this.removeNames_(zoomStatus.geneNames, names);
+        break;
+      case 'setCondition':
+        // Totally replace the regex.
+        zoomStatus.conditionNames = names;
+        break;
+      case 'addCondition':
+        // Concat the two names.
+        this.removeNames_(zoomStatus.conditionNames, names);
+        names.forEach(function(name) {
+          zoomStatus.conditionNames.push(name);
+        });
+        break;
+      case 'removeCondition':
+        // Remove the names.
+        this.removeNames_(zoomStatus.conditionNames, names);
+        break;
+    }
+  }.bind(this));
+  var zoomStatus = this.data.zoomStack.pop();
+  this.load(zoomStatus.matrixName, zoomStatus.dataName, zoomStatus.geneNames,
+    zoomStatus.conditionNames);
+};
+
+/**
+ * Remove genes or conditions from previous status in expression.
+ * @param {!Array<string>} originalNames Original names from previous zoom
+ *      status.
+ * @param {!Array<string>} removeNames Names need to be removed.
+ * @private
+ */
+genotet.ExpressionLoader.prototype.removeNames_ = function(originalNames,
+                                                           removeNames) {
+  removeNames.forEach(function(name) {
+    var geneIndex = originalNames.indexOf(name);
```

Comment:
I think it is unnecessary. Because originalNames is an array and the time complexity of this line is O(n). However, the time complexity of building a dict for an array is also O(n). Therefore, I think it is not that necessary to use DS here.